src/fw-update/fw-update-device.cpp

```diff
@@ -0,0 +1,262 @@
+        auto start = std::chrono::system_clock::now();
+        do {
+            dfu_status_payload status;
+            uint32_t transferred = 0;
+            auto sts = messenger->control_transfer(0xa1 /*DFU_GETSTATUS_PACKET*/, RS2_DFU_GET_STATUS, 0, 0, (uint8_t*)&status, sizeof(status), transferred, 5000);
+
+            if (status.is_in_state(state)) {
+                return true;
+            }
+
+            //test for dfu error state
+            if (status.is_error_state()) {
+                return false;
+            }
+
+            // FW doesn't set the bwPollTimeout value, therefore it is wrong to use status.bwPollTimeout
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+            auto curr = std::chrono::system_clock::now();
+            elapsed_milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(curr - start);
+        } while (elapsed_milliseconds < std::chrono::milliseconds(timeout));
+
+        return false;
+    }
+
+    update_device::update_device(const std::shared_ptr<context>& ctx, bool register_device_notifications, std::shared_ptr<platform::usb_device> usb_device)
+        : _context(ctx), _usb_device(usb_device)
+    {
+        auto messenger = _usb_device->open();
+
+        auto state = get_dfu_state(messenger);
+        detach(messenger);
+        state = get_dfu_state(messenger);
+
+        if (state != RS2_DFU_STATE_DFU_IDLE)
+            throw std::runtime_error("Cannot open recovery device!");
+
+        dfu_fw_status_payload paylaod;
+        uint32_t transferred = 0;
+        auto sts = messenger->control_transfer(0xa1, RS2_DFU_UPLOAD, 0, 0, reinterpret_cast<uint8_t*>(&paylaod), sizeof(paylaod), transferred, 10);
+        if (sts != platform::RS2_USB_STATUS_SUCCESS)
+            throw std::runtime_error("Failed to read serial number from recovery device!");
+
+        std::stringstream formattedBuffer;
+        for (auto i = 0; i < sizeof(paylaod.serial_number.serial); i++)
+            formattedBuffer << std::setfill('0') << std::setw(2) << std::hex << static_cast<int>(paylaod.serial_number.serial[i]);
+
+        _asic_serial_number = formattedBuffer.str();
+    }
+
+    update_device::~update_device()
+    {
+
+    }
+
+    //since we currently don't have serial number that we can use to identify the device
+    //this method looks for changes in the DFU device list
+    bool update_device::wait_for_device(int mask, uint32_t timeout) const
+    {
+        auto dfu_dev_count = _context->query_devices(mask).size();
+        if (dfu_dev_count == 0)
+            return true;
+
+        auto begin = std::chrono::system_clock::now();
+        while (dfu_dev_count == _context->query_devices(mask).size())
+        {
+            auto end = std::chrono::system_clock::now();
+            auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - begin).count();
+            if (duration > timeout)
+                break;
```

Comment:
Should probably be added to log - the anticipated device didn't resurface in DFU list